src/models/document/document.ts

```diff
@@ -68,6 +70,32 @@ export const DocumentModel = types
 
     setTileComments(tileId: string, comments: TileCommentsModelType) {
       self.comments.set(tileId, comments);
+    },
+
+    setUserStar(star: UserStarModelType) {
+      if (!self.stars.find( docStar => docStar.uid === star.uid )) {
+        self.stars.push(star);
+      }
+    },
+
+    updateUserStar(newStar: UserStarModelType) {
+      const starIndex = self.stars.findIndex(star => star.uid === newStar.uid);
+      if (starIndex >= 0) {
+        self.stars[starIndex] = newStar;
+      } else {
+        this.setUserStar(newStar);
```

Comment:
Are we sure this works the way you expect? I didn't think one could rely on `this` inside MST actions. We've generally had to rearrange things into multiple action blocks or use functions in an internal closure for actions to be able to make use of other actions.

Reply:
From the MST README:
>You cannot use this inside actions. Instead, use self. This makes it safe to pass actions around without binding them or wrapping them in arrow functions.

Reply:
Works in practice, but maybe I'm just getting lucky that `this` references the context that I'm hoping for.  I can always just write the code explicitly.  It's just two lines.

Reply:
In fact, because I've already done the check for existence of the uid in the star array, it's just one line.  Not really necessary to try to call that action which includes the uid check again.

Reply:
In fact, because I've already done the check for existence of the uid in the star array, it's just one line.  Not really necessary to try to call that action which includes the uid check again.

Reply:
Anyway, I removed the reference.  Cleaner not to call the function anyway.